_posts/2021-04-16-rustc-ios-and-an-m1.md

```diff
@@ -0,0 +1,177 @@
+It's as easy as:
+
+```
+arch -x86_64 $SHELL
+make build-kotlin
+```
+
+At least if you have Java set up correctly...
+The default Android emulator isn't usable on M1 hardware yet, but Google is working on a compatible one: [Android M1 emulator preview][android-emulator].
+It's usable enough for some testing, but for that part I most often switch back to my Linux Desktop (that has the additional CPU power on top).
+
+## Glean iOS on an M1
+
+Now we're getting to the interesting part: Native iOS development on an M1.
+Obviously for Apple this is a priority:
+Their new machines should become the main machine people do iOS development on.
+Thus Xcode gained `aarch64` support in version 12 long before the hardware was available.
+That caused quite some issues with existing tooling, such as the [dependency manager Carthage][carthage silicon issue].
+Here's the issue:
+
+* When compiling for iOS hardware you would pick a target named `aarch64-apple-ios`,
+  because ... iPhones and iPads are ARM-based since forever.
+* When compiling for the iOS simulator you would pick a target named `x86_64-apple-ios`,
+  because conveniently the simulator uses the host's CPU (that's what makes it fast)
+
+So when the compiler saw `x86_64` and `iOS` it knew "Ah, simulator target"
+and when it saw `aarch64` and `ios` it knew "Ah, hardware".
+And everyone went with this, Xcode happily built both targets and, if asked to, was able to bundle them into one package.
+
+With the introduction of Apple Silicion[^3]
+the iOS simulator run on these machines would _also_ be `aarch64`[^4],
+and also contain `ios`, but not be for the iOS hardware.
+
+Now Xcode and the compiler will get confused what to put where when building on M1 hardware for both iOS hardware and the host architecture.
+
+So the compiler toolchain gained knowledge of a new thing: `arm64-apple-ios14.0-simulator`,
+explicitly marking the simulator target.
+The compiler knows from where to pick the libraries and other SDK files when using that target.
+You still can't put code compiled for `arm64-apple-ios` and `arm64-apple-ios14.0-simulator` into the same universal binary[^5],
+because you can have each architecture only once (the `arm64` part in there).
+That's what Carthage and others stumbled over.
+
+Again Apple prepared for that and for a long time they want you to use [XCFramework bundles][XCFramework bundles][^6].
+Carthage just didn't used to support that.
+The [0.37.0 release][carthage release] fixed that.
+
+That still leaves Rust behind, as it doesn't know the new `-simulator` target.
+But as always the Rust community is ahead of the game and [deg4uss3r] started adding a new target in [Rust PR #81966][rust sim target].
+He got half way there when I jumped in to push it over the finish line.
+How these targets work and how LLVM picks the right things to put into the compiled artifacts is severly underdocumented,
```

Comment:
Is there work to improve the documentation?

Reply:
I didn't look much further, LLVM is _huge_ and I have no time to get familiar with their stuff.